File: c++_code/34.在排序数组中查找元素的第一个和最后一个位置.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */

// @lc code=start
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
          if(nums.size()==0) return {-1,-1};
          int first=findfirst(nums, target);
          int last=findlast(nums, target)-1;
          if(first==nums.size()||nums[first]!=target) {return {-1,-1};}
          return {first, last};

    }
    int findfirst(vector<int>& nums,int target) {
        //找第一个大于等于target的元素
        //有可能返回大于目标值的元素
        int l=0,r=nums.size();
        while(l<r) {
            int mid=l+(r-l)/2;
            if(nums[mid]>=target)
            {
                //如果nums[mid]>=target,则目标位置一定在mid处或mid左边
                r=mid;
            }
            else if(nums[mid]<target)
            {
                //若nums[mid]<target,则目标位置一定在mid右侧
                l=mid+1;
            }
          
        }
        return l;
    }

    int findlast(vector<int>& nums,int target) {
        //找第一个大于target的元素
        int l=0,r=nums.size();
        while(l<r) {
            int mid=l+(r-l)/2;
            if(nums[mid]>target)
            {
                //如果nums[mid]>target,则目标位置一定在mid处或mid左边
                r=mid;
            }
            else if(nums[mid]<=target)
            {
                //若nums[mid]<target,则目标位置一定在mid右侧
                l=mid+1;
            }
          
        }
        return r;
    }
    
};
// @lc code=end

File: js_code/专项/二分查找/34. 在排序数组中查找元素的第一个和最后一个位置.js
`
给定一个按照升序排列的整数数组 nums，和一个目标值 target。找出给定目标值在数组中的开始位置和结束位置。

如果数组中不存在目标值 target，返回 [-1, -1]。
`

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
 var searchRange = function(nums, target) {
  let l=0,r=nums.length-1
  while(l<=r){
      if(nums[l]!==target) l++;
      if(nums[r]!==target) r--
      if(nums[l]===target&&nums[r]===target) break
  }
  if(l>r) return[-1,-1]
  return [l,r]

};